ggu/main.c

```diff
@@ -16,11 +16,313 @@
+	num_policies = data[0];
+	data = data + 1;
+	proto_ptr = (uint16_t *)data;
+	policy_ptr = (uint8_t *)((uint16_t *)data + num_policies);
+
+	/* Loop over each policy decision on the packet. */
+	for (j = 0; j < num_policies; j++) {
+		int ret;
+		struct ggu_policy policy;
+
+		ret = policy_parse_next(proto_ptr, policy_ptr, &policy);
+		if (ret < 0)
+			break;
+
+		process_single_policy(&policy, ggu_conf);
+	}
+
+free_packet:
+	rte_pktmbuf_free(pkt);
+}
+
+static int
+ggu_proc(void *arg)
+{
+	int ret;
+	uint32_t lcore = rte_lcore_id();
+	struct ggu_config *ggu_conf = (struct ggu_config *)arg;
+	uint8_t port_in = ggu_conf->net->front.id;
+	uint16_t rx_queue = ggu_conf->rx_queue_front;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: the GK-GT unit is running at lcore = %u\n", lcore);
+
+	/* Wait for network devices to start. */
+	while (ggu_conf->net->configuring)
+		;
+
+	while (ggu_conf->net->front.rss_configured == false)
+		;
+
+	/* We can get the RSS configuration only after it's configured. */
+	ret = gatekeeper_get_rss_config(port_in, &ggu_conf->rss_conf);
+	if (ret < 0)
+		exiting = true;
+
+	/*
+	 * Setup the ntuple filter that assign the GK-GT packets
+	 * to its queue for both IPv4 and IPv6 addresses.
+	 * The ntuple Filters should be configured
+	 * after the network devices are started.
+	 */
+	ret = ntuple_filter_add(port_in,
+			rte_cpu_to_be_16(GGU_SRC_PORT),
+			rte_cpu_to_be_16(GGU_DST_PORT),
+			ggu_conf->rx_queue_front);
+	if (ret < 0)
+		exiting = true;
+
+	while (likely(!exiting)) {
+		uint16_t i;
+		uint16_t num_rx;
+		struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+
+		/* Load a set of GK-GT packets from the back NIC. */
+		num_rx = rte_eth_rx_burst(port_in, rx_queue, bufs,
+				GATEKEEPER_MAX_PKT_BURST);
+		if (unlikely(num_rx == 0))
+			continue;
+
+		for (i = 0; i < num_rx; i++) {
+			process_single_packet(bufs[i], ggu_conf);
+		}
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"ggu: the GK-GT unit at lcore = %u is exiting\n", lcore);
+
+	return cleanup_ggu(ggu_conf);
+}
+
+int
+run_ggu(struct net_config *net_conf, struct ggu_config *ggu_conf)
+{
+	int ret;
+
+	if (ggu_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (net_conf == NULL) {
+		ret = -1;
+		goto configuration;
+	}
+
+	ggu_conf->net = net_conf;
+
+	ret = get_queue_id(&ggu_conf->net->front,
+		QUEUE_TYPE_RX, ggu_conf->lcore_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "gk: cannot assign an RX queue for the front interface for lcore %u\n",
+			ggu_conf->lcore_id);
+		goto configuration;
+	}
+	ggu_conf->rx_queue_front = (uint16_t)ret;
+
+	ret = rte_eal_remote_launch(ggu_proc, ggu_conf, ggu_conf->lcore_id);
+	if (ret) {
+		RTE_LOG(ERR, EAL, "lcore %u failed to launch GGU\n",
+			ggu_conf->lcore_id);
+		ret = -1;
+		goto configuration;
+	}
+
+	ret = 0;
+	goto out;
+
+configuration:
+	rte_free(ggu_conf);
+	ggu_conf = NULL;
+out:
+	return ret;
+}
+
+struct ggu_config *
+alloc_ggu_conf(void)
+{
+	return rte_calloc("ggu_config", 1, sizeof(struct ggu_config), 0);
```

Comment:
It should be dynamic because there may not be a GK-GT unit, for example on a Grantor server. But it should return an error if trying to allocate the second instance. The network isn't an equivalent example because there'll alway be a network configuration.